From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Sun, 23 Aug 2020 13:11:40 +0200
Subject: [PATCH 1/1] Implement emscripten_num_logical_cores for Node.js

Since navigator['hardwareConcurrency'] can't be used on Node.js.

Upstream-Status: Accepted [https://github.com/emscripten-core/emscripten/commit/0e57ea0d95f081c9bdb45a77d938996cf0e723d2]

diff --git a/src/library_pthread.js b/src/library_pthread.js
index 1111111..2222222 100644
--- a/src/library_pthread.js
+++ b/src/library_pthread.js
@@ -561,6 +561,9 @@ var LibraryPThread = {
   },

   emscripten_num_logical_cores: function() {
+#if ENVIRONMENT_MAY_BE_NODE
+    if (ENVIRONMENT_IS_NODE) return require('os').cpus().length;
+#endif
     return navigator['hardwareConcurrency'];
   },

