From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Thu, 13 Aug 2020 15:10:00 +0200
Subject: [PATCH 1/1] Remove "Blocking on the main thread is very dangerous"
 warning

Upstream-Status: Inappropriate [disable feature]
This patch is just for our convenience.

diff --git a/system/lib/libc/musl/src/thread/pthread_cond_timedwait.c b/system/lib/libc/musl/src/thread/pthread_cond_timedwait.c
index 1111111..2222222 100644
--- a/system/lib/libc/musl/src/thread/pthread_cond_timedwait.c
+++ b/system/lib/libc/musl/src/thread/pthread_cond_timedwait.c
@@ -80,13 +80,6 @@ int __pthread_cond_timedwait(pthread_cond_t *restrict c, pthread_mutex_t *restri
 	int e, seq, clock = c->_c_clock, cs, shared=0, oldstate, tmp;
 	volatile int *fut;
 
-#ifdef __EMSCRIPTEN__
-	// TODO: Optimize this away in MINIMAL_RUNTIME.
-	if (pthread_self() == emscripten_main_browser_thread_id()) {
-		emscripten_check_blocking_allowed();
-	}
-#endif
-
 	if ((m->_m_type&15) && (m->_m_lock&INT_MAX) != __pthread_self()->tid)
 		return EPERM;
 
