From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Sat, 14 Aug 2021 21:00:00 +0200
Subject: [PATCH 1/1] Add missing Closure extern for ftruncateSync

Upstream-Status: Accepted [https://github.com/emscripten-core/emscripten/commit/0fe9b0b2f169a565fb5f419e696ddaeb710de8ef]

diff --git a/third_party/closure-compiler/node-externs/fs.js b/third_party/closure-compiler/node-externs/fs.js
index 1111111..2222222 100644
--- a/third_party/closure-compiler/node-externs/fs.js
+++ b/third_party/closure-compiler/node-externs/fs.js
@@ -56,6 +56,12 @@ fs.truncate = function(fd, len, callback) {};
  */
 fs.truncateSync = function(fd, len) {};
 
+/**
+ * @param {*} fd
+ * @param {number} len
+ */
+fs.ftruncateSync = function(fd, len) {};
+
 /**
  * @param {string} path
  * @param {number} uid
