From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Sat, 25 Jun 2022 14:40:00 +0200
Subject: [PATCH 1/1] [WasmFS] Allow non-private mmaps

Upstream-Status: Pending

diff --git a/system/lib/wasmfs/syscalls.cpp b/system/lib/wasmfs/syscalls.cpp
index 1111111..2222222 100644
--- a/system/lib/wasmfs/syscalls.cpp
+++ b/system/lib/wasmfs/syscalls.cpp
@@ -1516,12 +1516,6 @@ intptr_t _mmap_js(
     return -ENODEV;
   }
 
-  // TODO: handle non-private mmaps. Those can be optimized in interesting ways
-  //       like avoiding an allocation and a copy as we do below (whereas a
-  //       private mmap is always a copy into a new, private region not shared
-  //       with anything else).
-  assert(flags & MAP_PRIVATE);
-
   // Align to a wasm page size, as we expect in the future to get wasm
   // primitives to do this work, and those would presumably be aligned to a page
   // size. Aligning now avoids confusion later.
