From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Sun, 27 Mar 2022 20:00:00 +0200
Subject: [PATCH 1/1] Fix absolute path handling on Windows

See: https://github.com/emscripten-core/emscripten/pull/16296#issuecomment-1079979886

Upstream-Status: Pending

diff --git a/src/library_path.js b/src/library_path.js
index 1111111..2222222 100644
--- a/src/library_path.js
+++ b/src/library_path.js
@@ -6,7 +6,14 @@
 
 mergeInto(LibraryManager.library, {
   $PATH: {
-    isAbs: (path) => path.charAt(0) === '/',
+    isAbs: (path) => {
+#if ENVIRONMENT_MAY_BE_NODE
+      if (ENVIRONMENT_IS_NODE) {
+        return nodePath['isAbsolute'](path);
+      } else
+#endif
+      return path.charAt(0) === '/';
+    },
     // split a filename into [root, dir, basename, ext], unix version
     // 'root' is just a slash, or nothing.
     splitPath: (filename) => {
