From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Sun, 26 Jun 2022 10:51:53 +0200
Subject: [PATCH 1/1] [WasmFS] Implement FS.unlink

Upstream-Status: Pending

diff --git a/emcc.py b/emcc.py
index 1111111..2222222 100644
--- a/emcc.py
+++ b/emcc.py
@@ -2092,6 +2092,7 @@ def phase_linker_setup(options, state, newargs, user_settings):
       settings.REQUIRED_EXPORTS += [
         '_wasmfs_write_file',
         '_wasmfs_mkdir',
+        '_wasmfs_unlink',
         '_wasmfs_chdir',
         '_wasmfs_symlink',
         '_wasmfs_chmod',
diff --git a/src/library_wasmfs.js b/src/library_wasmfs.js
index 1111111..2222222 100644
--- a/src/library_wasmfs.js
+++ b/src/library_wasmfs.js
@@ -124,7 +124,12 @@ mergeInto(LibraryManager.library, {
     // TODO: open
     // TODO: create
     // TODO: close
-    // TODO: unlink
+    unlink: (path) => {
+      return withStackSave(() => {
+        var pathBuffer = allocateUTF8OnStack(path);
+        return __wasmfs_unlink(pathBuffer);
+      });
+    },
     chdir: (path) => {
       return withStackSave(() => {
         var buffer = allocateUTF8OnStack(path);
diff --git a/system/lib/wasmfs/js_api.cpp b/system/lib/wasmfs/js_api.cpp
index 1111111..2222222 100644
--- a/system/lib/wasmfs/js_api.cpp
+++ b/system/lib/wasmfs/js_api.cpp
@@ -100,6 +100,10 @@ int _wasmfs_mkdir(char* path, int mode) {
   return __syscall_mkdirat(AT_FDCWD, (intptr_t)path, mode);
 }
 
+int _wasmfs_unlink(char* path) {
+  return __syscall_unlinkat(AT_FDCWD, (intptr_t)path, 0);
+}
+
 int _wasmfs_chdir(char* path) { return __syscall_chdir((intptr_t)path); }
 
 int _wasmfs_symlink(char* old_path, char* new_path) {
