From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Wed, 25 Aug 2021 10:50:00 +0200
Subject: [PATCH 1/1] Enable SIMD specific code on WebAssembly

Upstream-Status: Pending

diff --git a/libimagequant.c b/libimagequant.c
index 1111111..2222222 100644
--- a/libimagequant.c
+++ b/libimagequant.c
@@ -160,7 +160,7 @@ static void liq_verbose_printf_flush(liq_attr *attr)
 #if USE_SSE
 inline static bool is_sse_available()
 {
-#if (defined(__x86_64__) || defined(__amd64))
+#if (defined(__x86_64__) || defined(__amd64) || defined(__EMSCRIPTEN__))
     return true;
 #else
     int a,b,c,d;
