From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Sun, 4 Sep 2022 17:09:21 +0200
Subject: [PATCH 1/1] Ensure blocked operations are knocked out of the cache

See: https://github.com/kleisauke/wasm-vips/pull/24#discussion_r962325956

Upstream-Status: Pending

diff --git a/libvips/iofuncs/cache.c b/libvips/iofuncs/cache.c
index 1111111..2222222 100644
--- a/libvips/iofuncs/cache.c
+++ b/libvips/iofuncs/cache.c
@@ -798,8 +798,10 @@ vips_cache_operation_lookup( VipsOperation *operation )
 	result = NULL;
 
 	if( (hit = g_hash_table_lookup( vips_cache_table, operation )) ) {
-		if( hit->invalid ) {
-			/* There but has been tagged for removal.
+		if( hit->invalid ||
+			(VIPS_OPERATION_GET_CLASS( hit->operation )->flags &
+			 VIPS_OPERATION_BLOCKED) ) {
+			/* Has been tagged for removal, or has been blocked.
 			 */
 			vips_cache_remove( hit->operation );
 			hit = NULL;
