From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Tue, 23 Aug 2022 10:45:02 +0200
Subject: [PATCH 1/1] Disable NLS

Upstream-Status: Inappropriate [disable feature]
This patch is just for our convenience.

diff --git a/meson.build b/meson.build
index 1111111..2222222 100644
--- a/meson.build
+++ b/meson.build
@@ -547,20 +547,6 @@ gettext_domain = 'vips@0@.@1@'.format(version_major, version_minor)
 cfg_var.set_quoted('GETTEXT_PACKAGE', gettext_domain)
 cfg_var.set_quoted('VIPS_PREFIX', prefix_dir)
 cfg_var.set_quoted('VIPS_LIBDIR', lib_dir)
-if cc.has_function('ngettext')
-    cfg_var.set('ENABLE_NLS', 1)
-    have_bind_textdomain_codeset = cc.has_function('bind_textdomain_codeset')
-else
-    libintl_dep = cc.find_library('intl', required: false)
-    if libintl_dep.found()
-        libvips_deps += libintl_dep
-        cfg_var.set('ENABLE_NLS', 1)
-        have_bind_textdomain_codeset = cc.has_function('bind_textdomain_codeset', prefix: '#include <libintl.h>', dependencies: libintl_dep)
-    else
-        have_bind_textdomain_codeset = false
-    endif
-endif
-cfg_var.set('HAVE_BIND_TEXTDOMAIN_CODESET', have_bind_textdomain_codeset)
 
 if host_os == 'darwin'
     profile_dir = '/Library/ColorSync/Profiles'
