From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Sun, 14 Aug 2022 12:12:00 +0200
Subject: [PATCH 1/1] Disable concurrency in jxl{load,save}

Upstream-Status: Inappropriate [Emscripten specific]

diff --git a/libvips/foreign/jxlload.c b/libvips/foreign/jxlload.c
index 1111111..2222222 100644
--- a/libvips/foreign/jxlload.c
+++ b/libvips/foreign/jxlload.c
@@ -157,7 +157,7 @@ vips_foreign_load_jxl_build( VipsObject *object )
 #endif /*DEBUG*/
 
 	jxl->runner = JxlThreadParallelRunnerCreate( NULL, 
-		vips_concurrency_get() );
+		0 );
 	jxl->decoder = JxlDecoderCreate( NULL );
 
 	if( JxlDecoderSubscribeEvents( jxl->decoder, 
diff --git a/libvips/foreign/jxlsave.c b/libvips/foreign/jxlsave.c
index 1111111..2222222 100644
--- a/libvips/foreign/jxlsave.c
+++ b/libvips/foreign/jxlsave.c
@@ -253,7 +253,7 @@ vips_foreign_save_jxl_build( VipsObject *object )
 		jxl->lossless = TRUE;
 
 	jxl->runner = JxlThreadParallelRunnerCreate( NULL, 
-		vips_concurrency_get() );
+		0 );
 	jxl->encoder = JxlEncoderCreate( NULL );
 
 	if( JxlEncoderSetParallelRunner( jxl->encoder, 
