From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: John Cupitt <jcupitt@gmail.com>
Date: Tue, 9 Feb 2021 16:07:51 +0000
Subject: [PATCH 1/1] don't put glib includes inside extern "C" blocks

see https://github.com/libvips/libvips/issues/1987

diff --git a/libvips/include/vips/dispatch.h b/libvips/include/vips/dispatch.h
index 1111111..2222222 100644
--- a/libvips/include/vips/dispatch.h
+++ b/libvips/include/vips/dispatch.h
@@ -33,11 +33,12 @@
 #ifndef IM_DISPATCH_H
 #define IM_DISPATCH_H
 
+#include <glib-object.h>
+
 #ifdef __cplusplus
 extern "C" {
 #endif /*__cplusplus*/
 
-#include <glib-object.h>
 #include <vips/vips.h>
 #include <vips/util.h>
 
diff --git a/libvips/include/vips/vips.h b/libvips/include/vips/vips.h
index 1111111..2222222 100644
--- a/libvips/include/vips/vips.h
+++ b/libvips/include/vips/vips.h
@@ -80,10 +80,6 @@
 #ifndef VIPS_VIPS_H
 #define VIPS_VIPS_H
 
-#ifdef __cplusplus
-extern "C" {
-#endif /*__cplusplus*/
-
 #include <glib.h>
 #include <glib/gstdio.h>
 #include <gmodule.h>
@@ -93,6 +89,10 @@ extern "C" {
  */
 #include <gio/gio.h>
 
+#ifdef __cplusplus
+extern "C" {
+#endif /*__cplusplus*/
+
 /* If we're being parsed by SWIG, remove gcc attributes.
  */
 #ifdef SWIG
