From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Mon, 22 Aug 2022 14:10:04 +0200
Subject: [PATCH 1/4] Support shared modules on Emscripten

Upstream-Status: Pending

diff --git a/libvips/meson.build b/libvips/meson.build
index 1111111..2222222 100644
--- a/libvips/meson.build
+++ b/libvips/meson.build
@@ -79,11 +79,25 @@ if ['darwin', 'ios'].contains(host_os)
   module_suffix = 'so'
 endif
 
+# For shared modules on Emscripten:
+# - Use .wasm as shared module suffix
+# - Compile and link with -sSIDE_MODULE=2
+# - Ensure g_module_check_init is not DCE'd
+module_c_args = ''
+module_link_args = ''
+if host_os == 'emscripten'
+    module_suffix = 'wasm'
+    module_c_args = '-sSIDE_MODULE=2'
+    module_link_args = [module_c_args, '-sEXPORTED_FUNCTIONS=_g_module_check_init']
+endif
+
 if magick_module
     shared_module('vips-magick',
         'module/magick.c',
         magick_module_sources,
         magick_module_headers,
+        c_args: module_c_args,
+        link_args: module_link_args,
         name_prefix: '',
         name_suffix: module_suffix,
         dependencies: [libvips_dep, magick_dep],
@@ -96,6 +110,8 @@ if libjxl_module
     shared_module('vips-jxl',
         'module/jxl.c',
         jpeg_xl_module_sources,
+        c_args: module_c_args,
+        link_args: module_link_args,
         name_prefix: '',
         name_suffix: module_suffix,
         dependencies: [libvips_dep, libjxl_dep, libjxl_threads_dep],
@@ -108,6 +124,8 @@ if libheif_module
     shared_module('vips-heif',
         'module/heif.c',
         heif_module_sources,
+        c_args: module_c_args,
+        link_args: module_link_args,
         name_prefix: '',
         name_suffix: module_suffix,
         dependencies: [libvips_dep, libheif_dep],
@@ -120,6 +138,8 @@ if libpoppler_module
     shared_module('vips-poppler',
         'module/poppler.c',
         poppler_module_sources,
+        c_args: module_c_args,
+        link_args: module_link_args,
         name_prefix: '',
         name_suffix: module_suffix,
         dependencies: [libvips_dep, libpoppler_dep, cairo_dep],
@@ -132,6 +152,8 @@ if openslide_module
     shared_module('vips-openslide',
         'module/openslide.c',
         openslide_module_sources,
+        c_args: module_c_args,
+        link_args: module_link_args,
         name_prefix: '',
         name_suffix: module_suffix,
         dependencies: [libvips_dep, openslide_dep],

From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Mon, 22 Aug 2022 19:33:06 +0200
Subject: [PATCH 2/4] Apply GNU visibility arguments to shared modules

Upstream-Status: Pending

diff --git a/libvips/meson.build b/libvips/meson.build
index 1111111..2222222 100644
--- a/libvips/meson.build
+++ b/libvips/meson.build
@@ -101,6 +101,7 @@ if magick_module
         name_prefix: '',
         name_suffix: module_suffix,
         dependencies: [libvips_dep, magick_dep],
+        gnu_symbol_visibility: 'hidden',
         install: true,
         install_dir: module_dir
     )
@@ -115,6 +116,7 @@ if libjxl_module
         name_prefix: '',
         name_suffix: module_suffix,
         dependencies: [libvips_dep, libjxl_dep, libjxl_threads_dep],
+        gnu_symbol_visibility: 'hidden',
         install: true,
         install_dir: module_dir
     )
@@ -129,6 +131,7 @@ if libheif_module
         name_prefix: '',
         name_suffix: module_suffix,
         dependencies: [libvips_dep, libheif_dep],
+        gnu_symbol_visibility: 'hidden',
         install: true,
         install_dir: module_dir
     )
@@ -143,6 +146,7 @@ if libpoppler_module
         name_prefix: '',
         name_suffix: module_suffix,
         dependencies: [libvips_dep, libpoppler_dep, cairo_dep],
+        gnu_symbol_visibility: 'hidden',
         install: true,
         install_dir: module_dir
     )
@@ -157,6 +161,7 @@ if openslide_module
         name_prefix: '',
         name_suffix: module_suffix,
         dependencies: [libvips_dep, openslide_dep],
+        gnu_symbol_visibility: 'hidden',
         install: true,
         install_dir: module_dir
     )
     
From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Mon, 22 Aug 2022 19:54:04 +0200
Subject: [PATCH 3/4] Ensure shared modules are not linked against transitive deps

`EM_JS` (as used by libffi) is not supported in side modules,
so try to avoid linking with that.

Upstream-Status: Inappropriate [Emscripten specific]

diff --git a/libvips/meson.build b/libvips/meson.build
index 1111111..2222222 100644
--- a/libvips/meson.build
+++ b/libvips/meson.build
@@ -30,8 +30,7 @@ libvips_lib = library('vips',
 )
 
 libvips_dep = declare_dependency(
-    link_with: libvips_lib,
-    dependencies: libvips_deps,
+    dependencies: [config_dep, libvips_headers_dep],
 )
 
 pkg.generate(
@@ -100,7 +99,7 @@ if magick_module
         link_args: module_link_args,
         name_prefix: '',
         name_suffix: module_suffix,
-        dependencies: [libvips_dep, magick_dep],
+        dependencies: [gmodule_dep, libvips_dep, magick_dep],
         gnu_symbol_visibility: 'hidden',
         install: true,
         install_dir: module_dir
@@ -115,7 +114,7 @@ if libjxl_module
         link_args: module_link_args,
         name_prefix: '',
         name_suffix: module_suffix,
-        dependencies: [libvips_dep, libjxl_dep, libjxl_threads_dep],
+        dependencies: [gmodule_dep, libvips_dep, libjxl_dep, libjxl_threads_dep],
         gnu_symbol_visibility: 'hidden',
         install: true,
         install_dir: module_dir
@@ -130,7 +129,7 @@ if libheif_module
         link_args: module_link_args,
         name_prefix: '',
         name_suffix: module_suffix,
-        dependencies: [libvips_dep, libheif_dep],
+        dependencies: [gmodule_dep, libvips_dep, libheif_dep],
         gnu_symbol_visibility: 'hidden',
         install: true,
         install_dir: module_dir
@@ -145,7 +144,7 @@ if libpoppler_module
         link_args: module_link_args,
         name_prefix: '',
         name_suffix: module_suffix,
-        dependencies: [libvips_dep, libpoppler_dep, cairo_dep],
+        dependencies: [gmodule_dep, libvips_dep, libpoppler_dep, cairo_dep],
         gnu_symbol_visibility: 'hidden',
         install: true,
         install_dir: module_dir
@@ -160,7 +159,7 @@ if openslide_module
         link_args: module_link_args,
         name_prefix: '',
         name_suffix: module_suffix,
-        dependencies: [libvips_dep, openslide_dep],
+        dependencies: [gmodule_dep, libvips_dep, openslide_dep],
         gnu_symbol_visibility: 'hidden',
         install: true,
         install_dir: module_dir

From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Sun, 28 Aug 2022 11:32:25 +0200
Subject: [PATCH 4/4] Load modules from Module.dynamicLibraries instead

Upstream-Status: Inappropriate [Emscripten specific]

diff --git a/libvips/iofuncs/init.c b/libvips/iofuncs/init.c
index 1111111..2222222 100644
--- a/libvips/iofuncs/init.c
+++ b/libvips/iofuncs/init.c
@@ -311,6 +311,43 @@ vips_load_plugins( const char *fmt, ... )
 
 	return( result );
 }
+
+#ifdef __EMSCRIPTEN__
+extern char **_emscripten_get_dynamic_libraries_js();
+
+static void
+vips_load_modules_wasm()
+{
+	char **libs, **p;
+
+	/* Do nothing if modules aren't supported.
+	 */
+	if( !g_module_supported() )
+		return;
+
+	libs = _emscripten_get_dynamic_libraries_js();
+
+	for( p = libs; *p; p++ ) {
+		GModule *module;
+
+		g_info( "loading \"%s\"", *p );
+
+		module = g_module_open( *p, G_MODULE_BIND_LAZY );
+		if( !module ) {
+			g_warning( _( "unable to load \"%s\" -- %s" ),
+				*p, g_module_error() );
+		}
+		g_free( *p );
+
+		/* Modules will almost certainly create new types, so
+		 * they can't be unloaded.
+		 */
+		g_module_make_resident( module );
+	}
+
+	g_free( libs );
+}
+#endif /*__EMSCRIPTEN__*/
 #endif /*ENABLE_MODULES*/
 
 /* Install this log handler to hide warning messages.
@@ -600,6 +637,11 @@ vips_init( const char *argv0 )
 	vips_g_input_stream_get_type(); 
 
 #ifdef ENABLE_MODULES
+#ifdef __EMSCRIPTEN__
+	/* Load any vips8 modules from the Module.dynamicLibraries array.
+	 */
+	vips_load_modules_wasm();
+#else /*!__EMSCRIPTEN__*/
 	/* Load any vips8 modules from the vips libdir. Keep going, even if
 	 * some modules fail to load. 
 	 *
@@ -609,6 +651,7 @@ vips_init( const char *argv0 )
 	 */
 	(void) vips_load_plugins( "%s/vips-modules-%d.%d", 
 		libdir, VIPS_MAJOR_VERSION, VIPS_MINOR_VERSION );
+#endif /*__EMSCRIPTEN__*/
 
 #if ENABLE_DEPRECATED
 	/* Load any vips8 plugins from the vips libdir.
